### PR TITLE
[rust 2021] migrate to rust 2021

### LINKED ISSUE #1149: [rust] try Rust 2021
**Summary**

https://blog.rust-lang.org/2021/07/21/Rust-2021-public-testing.html

### PR DESCRIPTION
**Pull request by PsiACE**
Signed-off-by: Chojan Shang <psiace@outlook.com>

I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

Migrate to rust 2021

cargo fmt fails with error "Invalid value for \`--edition\`", see:
  - https://github.com/rust-lang/rust/issues/87363
wait: 
  - https://github.com/rust-lang/rust/pull/87468

## Changelog

- Improvement
- Build/Testing/CI

## Related Issues

Fixes #1149 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/datablocks/src/kernels/data_block_sort.rs [hunk 1]
```diff
@@ -87,7 +87,10 @@ impl DataBlock {
             .columns()
             .iter()
             .zip(rhs.columns().iter())
-            .map(|(a, b)| DataColumnCommon::merge_columns(a, b, indices))
+            .map(|(a, b)| {
+                let _ = &indices;
```
Comment (BohuTANG) [inline common/datablocks/src/kernels/data_block_sort.rs:91]: This line used for?

File: common/datavalues/src/arrays/kernels/take.rs [hunk 2]
```diff
@@ -141,9 +148,12 @@ pub unsafe fn take_no_null_primitive_opt_iter_unchecked<
 ) -> Arc<PrimitiveArray<T>> {
     let array_values = arr.values();
 
-    let iter = indices
-        .into_iter()
-        .map(|opt_idx| opt_idx.map(|idx| *array_values.get_unchecked(idx)));
+    let iter = indices.into_iter().map(|opt_idx| {
+        opt_idx.map(|idx| {
+            let _ = &array_values;
```
Comment (sundy-li) [inline common/datavalues/src/arrays/kernels/take.rs:153]: Why rust 2021 turns to be more complex.

File: common/datablocks/src/kernels/data_block_sort.rs (hunk 1 above)
Comment (PsiACE) [reply common/datablocks/src/kernels/data_block_sort.rs:91]: > This line used for?

see https://doc.rust-lang.org/nightly/edition-guide/rust-2021/disjoint-capture-in-closures.html

Auto gen by `cargo fix`. This is a conservative analysis: in many cases, these dummy lets can be safely removed and your program will work fine.

File: common/datavalues/src/arrays/kernels/take.rs (hunk 2 above)
Comment (BohuTANG) [reply common/datavalues/src/arrays/kernels/take.rs:153]: :sob:

Comment (PsiACE) [reply common/datavalues/src/arrays/kernels/take.rs:153]: > Why rust 2021 turns to be more complicated.

I don't think it becomes more complicated. 

> This is a conservative analysis: in many cases, these dummy lets can be safely removed and your program will work fine.

> Starting in Rust 2021, closures captures are more precise. Typically they will only capture the fields they use (in some cases, they might capture more than just what they use, see the Rust reference for full details).

**Comment by PsiACE** [comment]
We can try to adjust these codes later. Just announcing that we are moving towards Rust 2021.

**Comment by ariesdevil** [comment]
Since the Rust official says:
> Note that Rust 2021 is still unstable, so you can expect bugs and other changes! We recommend migrating your crates in a temporary copy of your code versus your main branch. If you do encounter problems, or find areas where quality could be improved (missing documentation, confusing error messages, etc) please file an issue and tell us about it! Thank you!

We'd better make this PR commit to an independent branch likes `rust-2021` instead of direct to master.
@PsiACE

**Comment by PsiACE** [comment]
> We'd better make this PR commit to an independent branch likes `rust-2021` instead of direct to master.

Okay, thanks.

File: common/datablocks/src/kernels/data_block_sort.rs (hunk 1 above)
Comment (zhang2014) [reply common/datablocks/src/kernels/data_block_sort.rs:91]: It doesn't look like a good style : (

Comment (zhang2014) [reply common/datablocks/src/kernels/data_block_sort.rs:91]: Maybe we can do it with macros? similar to '[&captures_var] (args) {} in C + +`.

Comment (PsiACE) [reply common/datablocks/src/kernels/data_block_sort.rs:91]: > Maybe we can do it with macros? similar to '[&captures_var] (args) {} in C++`.

We can try to remove these dummy let manually. If it is really necessary, consider other options.

**Comment by BohuTANG** [comment]
> Since the Rust official says:
> 
> > Note that Rust 2021 is still unstable, so you can expect bugs and other changes! We recommend migrating your crates in a temporary copy of your code versus your main branch. If you do encounter problems, or find areas where quality could be improved (missing documentation, confusing error messages, etc) please file an issue and tell us about it! Thank you!
> 
> We'd better make this PR commit to an independent branch likes `rust-2021` instead of direct to master.
> @PsiACE

If there are no major problems, merging to the main branch is ok, we also want to help rust 2021 mature faster, not simply testing, and 2021 will soon be released nightly version.
cc @PsiACE

**Comment by PsiACE** [comment]
> If there are no major problems, merging to the maater branch is ok, we also want to help Rust 2021 mature faster, not simply testing, and Rust 2021 nightly will be released soon.

There seems to be no problem, I will check if there is some room for improvement, and then merge before next Wednesday.

**Comment by PsiACE** [comment]
Finally, I removed all useless dummy lets. I think we can announce a successful migration to Rust 2021. The only regret is that the `cargo format` is temporarily invalid.

**Comment by BohuTANG** [comment]
> Finally, I removed all useless dummy lets. I think we can announce a successful migration to Rust 2021. The only regret is that the `cargo format` is temporarily invalid.

Great, waiting for https://github.com/rust-lang/rust/pull/87468 to merge.

**Comment by BohuTANG** [comment]
> > Finally, I removed all useless dummy lets. I think we can announce a successful migration to Rust 2021. The only regret is that the `cargo format` is temporarily invalid.
> 
> Great, waiting for [rust-lang/rust#87468](https://github.com/rust-lang/rust/pull/87468) to merge.

87468 has merged, we could start this task again.
cc @PsiACE

**Comment by PsiACE** [comment]
> 87468 has merged, we could start this task again.
> cc @PsiACE

Yes, but we still need to wait for the new nightly version. This PR may be completed later today.

**Comment by PsiACE** [comment]
@BohuTANG Done. Once merged, we will enter the era of Rust 2021.

**Review by BohuTANG — APPROVED** [review]
Looks great to me!

**Pull request merged**